git-branchless/src/revset/grammar.lalrpop

```diff
@@ -1,110 +1,97 @@
-    <a:Expr2> <op:BinOp1> <b:Expr15> => Expr::Fn(op.into(), vec![a, b]),
-    Expr15,
-};
-
-BinOp1: &'static str = {
-    ":" => "range",
-    ".." => "range",
-};
-
-Expr15: Expr = {
-    <pre:(<Prefix>)*> <e:Expr1> => {
-        let mut e = e;
-        for v in pre { e = Expr::Fn(v.into(), vec![e]); }
-        e
-    },
+// Below implements a hierarchy of operator precedences. The lower-numbered
+// `Expr`s bind less tightly than the higher-numbered `Expr`s.
+
+pub Expr: Expr<'input> = {
+    "(" <Expr> ")",
+    <lhs:Expr> "|"  <rhs:Expr2> => Expr::FunctionCall(Cow::Borrowed("union"), vec![lhs, rhs]),
+    <lhs:Expr> "+"  <rhs:Expr2> => Expr::FunctionCall(Cow::Borrowed("union"), vec![lhs, rhs]),
+    <lhs:Expr> "or" <rhs:Expr2> => Expr::FunctionCall(Cow::Borrowed("union"), vec![lhs, rhs]),
+    <Expr2>,
 }
 
-Expr1: Expr = {
-    <e:Expr0> <post:(<Postfix>)*> => {
-        let mut e = e;
-        for v in post { e = Expr::Fn(v.into(), vec![e]); }
-        e
-    },
+Expr2: Expr<'input> = {
+    <lhs:Expr2> "&"   <rhs:Expr3> => Expr::FunctionCall(Cow::Borrowed("intersection"), vec![lhs, rhs]),
+    <lhs:Expr2> "and" <rhs:Expr3> => Expr::FunctionCall(Cow::Borrowed("intersection"), vec![lhs, rhs]),
+    <lhs:Expr2> "-"   <rhs:Expr3> => Expr::FunctionCall(Cow::Borrowed("difference"),   vec![lhs, rhs]),
+    <lhs:Expr2> "%"   <rhs:Expr3> => Expr::FunctionCall(Cow::Borrowed("only"),         vec![lhs, rhs]),
+    <Expr3>
 }
 
-Expr0: Expr = {
-    Symbol2 => Expr::Name(<>),
-    <f:Symbol1> "(" <args:(Expr ",")*> <last:Expr?> ")" => {
-        // Function call.
-        let mut arg_list: Vec<Expr> = args.into_iter().map(|(e, _)| e).collect();
-        if let Some(last_arg) = last { arg_list.push(last_arg); }
-        Expr::Fn(f.into(), arg_list)
-    },
-    "(" <Expr> ")"
-};
+Expr3: Expr<'input> = {
+    <lhs:Expr3> ":"  <rhs:Expr4> =>  Expr::FunctionCall(Cow::Borrowed("range"),       vec![lhs, rhs]),
+    <lhs:Expr3> ":"              =>  Expr::FunctionCall(Cow::Borrowed("descendants"), vec![lhs,    ]),
+                ":"  <rhs:Expr4> =>  Expr::FunctionCall(Cow::Borrowed("ancestors"),   vec![     rhs]),
 
-Prefix: &'static str = {
-    // negation unimplemented for now
-    // "!" => "negate",
-    // "not " => "negate",
+    // For Mercurial users' familiarity.
+    <lhs:Expr3> "::"  <rhs:Expr4> =>  Expr::FunctionCall(Cow::Borrowed("range"),       vec![lhs, rhs]),
+    <lhs:Expr3> "::"              =>  Expr::FunctionCall(Cow::Borrowed("descendants"), vec![lhs,    ]),
+                "::"  <rhs:Expr4> =>  Expr::FunctionCall(Cow::Borrowed("ancestors"),   vec![     rhs]),
 
-    "::" => "ancestors",
+    <lhs:Expr3> ".." <rhs:Expr4> =>  Expr::FunctionCall(Cow::Borrowed("only"), vec![lhs, rhs]),
+    <lhs:Expr3> ".."             =>  Expr::FunctionCall(Cow::Borrowed("only"), vec![lhs, Expr::Name(Cow::Borrowed("."))]),
+                ".." <rhs:Expr4> =>  Expr::FunctionCall(Cow::Borrowed("only"), vec![Expr::Name(Cow::Borrowed(".")), rhs]),
```

Comment:
Would you consider reversing the `lhs, rhs` (ie `vec![rhs, lhs]`)  so that `foo..bar` would mean `only(bar, foo)`? This would allow revset's evaluation of `..` to match (as I understand it) how `git` handles double dot notation, [which is a pretty common notation in standard `git`](https://www.git-scm.com/docs/gitrevisions#_dotted_range_notations). 

To illustrate my concern, here's some example output from this branch:
```
❯ git co arxanas/revset
branchless: creating working copy snapshot
branchless: running command: git checkout arxanas/revset
Already on 'arxanas/revset'
Your branch is up to date with 'origin/arxanas/revset'.
branchless: processing checkout
⋮
◇ 0f06699 16h (master, remote origin/master) build: run `cargo upgrade`
┃
◯ 7e512e6 15h build(revset): build parser as part of `build.rs`
┃
◯ 87bc6fd 15h tests(eval): create `eval_and_sort` function
┃
◯ 96b891e 15h feat(revset): add `not` function
┃
◯ c9ec768 15h feat(revset): produce friendlier error messages
┃
◯ 3793bfd 15h refactor(revset): rewrite grammar to support `:` as a mixfix operator
┃
● bc0c8b7 15h (ᐅ arxanas/revset) feat(smartlog): support passing an arbitrary revset expression

❯ git rev-list master..@
bc0c8b750f514883c8abca09246fa72637b9be08
3793bfdb721ccb2c6e0b2bbed0bd36ceced3174c
c9ec76842ca33839b9770fb98f3f51185594e8cc
96b891edb5ab4dbda1154ea6f22c38317a518598
87bc6fdc6dd3fb50dc05e5b11c81067adea502f4
7e512e620b19925129669b5f11538c40b09c319e

❯ cargo run -- query master..@
    Finished dev [unoptimized] target(s) in 0.20s
     Running `target/debug/git-branchless query 'master..@'`
# no output

❯ cargo run -- query @..master
    Finished dev [unoptimized] target(s) in 0.90s
     Running `target/debug/git-branchless query '@..master'`
bc0c8b750f514883c8abca09246fa72637b9be08
3793bfdb721ccb2c6e0b2bbed0bd36ceced3174c
c9ec76842ca33839b9770fb98f3f51185594e8cc
96b891edb5ab4dbda1154ea6f22c38317a518598
87bc6fdc6dd3fb50dc05e5b11c81067adea502f4
7e512e620b19925129669b5f11538c40b09c319e

❯ git rev-list @..master
# no output

```

Reply:
Oops, thanks for catching! Fixed.